File: sdks/typescript/test/webhooks.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import {
  CregisClientError,
  CregisPaymentCallbackHandler,
  CregisWaasCallbackHandler,
  signProjectParameters,
} from "../src/index.js";

const API_KEY = "callback-api-key";

function signedBody(payload: Record<string, unknown>, apiKey = API_KEY): string {
  return JSON.stringify({ ...payload, sign: signProjectParameters(payload, apiKey) });
}

function commonEnvelope(): Record<string, unknown> {
  return {
    pid: 1382528827416576,
    nonce: "m8jisx",
    timestamp: 1687848653294,
  };
}

function waasBase(): Record<string, unknown> {
  return {
    ...commonEnvelope(),
    cid: 1382813146816512,
    chain_id: "195",
    token_id: "195",
    currency: "195@195",
    amount: "10.5",
  };
}

function paymentData(eventType: string): Record<string, unknown> {
  const common = {
    cregis_id: "po-1",
    order_id: "merchant-1",
    status: eventType === "expired" ? "expired" : eventType === "refunded" ? "canceled" : "paid",
  };
  if (eventType === "expired") return common;
  return {
    ...common,
    payment_address: "T-payment",
    receive_amount: "10",
    receive_currency: "USD",
    pay_amount: "10",
    pay_currency: "USDT-TRC20",
    exchange_rate: "1",
    tx_id: "tx-1",
    transact_time: 1719994383015,
  };
}

test("Payment callback verifies signatures and preserves event-specific wire types", () => {
  const handler = new CregisPaymentCallbackHandler(API_KEY);
  const notification = handler.verifyAndParse(signedBody({
    ...commonEnvelope(),
    event_name: "order",
    event_type: "refunded",
    data: {
      ...paymentData("refunded"),
      type: 1,
      refund_id: "rf-1",
      refund_status: 1,
      refund_created_time: 1719994183015,
      refund_transact_time: 1719994383015,
      future_field: "ignored",
    },
  }));

  assert.equal(notification.event_type, "refunded");
  if (notification.event_type === "refunded") {
    assert.equal(notification.data.refund_id, "rf-1");
    assert.equal(notification.data.refund_status, 1);
    assert.equal(notification.data.refund_transact_time, 1719994383015);
  }
  assert.equal(CregisPaymentCallbackHandler.CALLBACK_SUCCESS, "success");
});

test("Payment callback dispatches every documented event", () => {
  const handler = new CregisPaymentCallbackHandler(API_KEY);
  for (const event_type of ["paid", "paid_partial", "paid_over", "expired", "refunded", "paid_remain"] as const) {
    const result = handler.verifyAndParse(signedBody({
      ...commonEnvelope(),
      event_name: "order",
      event_type,
      data: paymentData(event_type),
    }));
    assert.equal(result.event_type, event_type);
  }
});

test("all four WaaS webhook contracts verify and parse", () => {
  const handler = new CregisWaasCallbackHandler(API_KEY);

  const deposit = handler.handleDepositCallback(signedBody({
    ...waasBase(),
    address: "deposit-address",
    status: "1",
    txid: "tx-1",
    block_time: "1734328473070",
  }));
  assert.equal(deposit.status, "1");

  const payout = handler.handlePayoutCallback(signedBody({
    ...waasBase(),
    address: "payout-address",
    third_party_id: "payout-1",
    status: 6,
    block_time: 1734328473070,
  }));
  assert.equal(payout.status, 6);

  const external = handler.handlePayoutExternalVerificationCallback(signedBody({
    ...waasBase(),
    third_party_id: "external-1",
    from_address: "from",
    to_address: "to",
  }));
  assert.equal(external.third_party_id, "external-1");

  const withdrawal = handler.handleWithdrawalCallback(signedBody({
    ...waasBase(),
    from_address: "from",
    to_address: "to",
    third_party_id: "withdrawal-1",
    status: 6,
  }));
  assert.equal(withdrawal.to_address, "to");
  assert.equal(CregisWaasCallbackHandler.CALLBACK_SUCCESS, "success");
  assert.equal(CregisWaasCallbackHandler.EXTERNAL_VERIFICATION_APPROVE, "ok");
  assert.equal(CregisWaasCallbackHandler.EXTERNAL_VERIFICATION_DENY, "deny");
});

test("webhooks reject bad signatures, missing fields, and wrong wire types", () => {
  const payment = new CregisPaymentCallbackHandler(API_KEY);
  const waas = new CregisWaasCallbackHandler(API_KEY);

  assert.throws(() => payment.verifyAndParse("{}"), CregisClientError);
  assert.throws(() => payment.verifyAndParse(signedBody({
    ...commonEnvelope(),
    event_name: "order",
    event_type: "future_event",
    data: {},
  })), /has no payload contract mapping/);
  assert.throws(() => payment.verifyAndParse(signedBody({
    ...commonEnvelope(),
    event_name: "order",
    event_type: "refunded",
    data: { ...paymentData("refunded"), transact_time: "1719994383015" },
  })), /transact_time must be integer/);
  assert.throws(() => payment.verifyAndParse(signedBody({
    ...commonEnvelope(),
    event_name: "order",
    event_type: "paid",
    data: {},
  })), /missing required field cregis_id/);
  assert.throws(() => waas.handleDepositCallback(signedBody({
    ...waasBase(),
    address: "deposit-address",
    status: 1,
    txid: "tx-1",
  })), /status must be string/);
});

test("stable nested Payment fixture remains compatible with Java SDK", () => {
  const rawBody = JSON.stringify({
    event_name: "order",
    event_type: "refunded",
    pid: 123456789,
    nonce: "abc123",
    timestamp: 1719994383015,
    data: {
      cregis_id: "po-test",
      order_id: "merchant-test",
      status: "canceled",
      payment_address: "T-payment",
      receive_amount: "10",
      receive_currency: "USD",
      pay_amount: "10",
      pay_currency: "USDT-TRC20",
      exchange_rate: "1",
      tx_id: "tx-test",
      transact_time: 1719994383015,
      refund_id: "rf-test",
      refund_status: 1,
    },
    sign: "79ecbe54381b0bd9b580fdca0fa536ac",
  });

  const result = new CregisPaymentCallbackHandler("fixture-api-key").verifyAndParse(rawBody);
  assert.equal(result.event_type, "refunded");
  if (result.event_type === "refunded") {
    assert.equal(result.data.refund_id, "rf-test");
  }
});

File: sdks/go/webhooks_test.go
package cregis

import (
	"encoding/json"
	"errors"
	"strings"
	"testing"
	"time"
)

const callbackAPIKey = "callback-api-key"

func callbackTestClock() time.Time { return time.UnixMilli(1687848653294) }

func signedCallbackBody(t *testing.T, payload map[string]any) []byte {
	t.Helper()
	signature, err := SignProjectParameters(payload, callbackAPIKey)
	if err != nil {
		t.Fatal(err)
	}
	signed := make(map[string]any, len(payload)+1)
	for key, value := range payload {
		signed[key] = value
	}
	signed["sign"] = signature
	body, err := json.Marshal(signed)
	if err != nil {
		t.Fatal(err)
	}
	return body
}

func callbackEnvelope() map[string]any {
	return map[string]any{
		"pid":       int64(1382528827416576),
		"nonce":     "m8jisx",
		"timestamp": int64(1687848653294),
	}
}

func paymentCallbackData(eventType string) map[string]any {
	status := "paid"
	if eventType == "expired" {
		status = "expired"
	} else if eventType == "refunded" {
		status = "canceled"
	}
	result := map[string]any{
		"cregis_id": "po-1",
		"order_id":  "merchant-1",
		"status":    status,
	}
	if eventType == "expired" {
		return result
	}
	result["payment_address"] = "T-payment"
	result["receive_amount"] = "10"
	result["receive_currency"] = "USD"
	result["pay_amount"] = "10"
	result["pay_currency"] = "USDT-TRC20"
	result["exchange_rate"] = "1"
	result["tx_id"] = "tx-1"
	result["transact_time"] = int64(1719994383015)
	return result
}

func waasCallbackBase() map[string]any {
	result := callbackEnvelope()
	result["cid"] = int64(1382813146816512)
	result["chain_id"] = "195"
	result["token_id"] = "195"
	result["currency"] = "195@195"
	result["amount"] = "10.5"
	return result
}

func cloneMap(source map[string]any) map[string]any {
	copy := make(map[string]any, len(source))
	for key, value := range source {
		copy[key] = value
	}
	return copy
}

func TestPaymentCallbackDispatchesAllDocumentedEvents(t *testing.T) {
	handler, err := NewPaymentCallbackHandler(callbackAPIKey, WithWebhookClock(callbackTestClock))
	if err != nil {
		t.Fatal(err)
	}
	for _, eventType := range []string{"paid", "paid_partial", "paid_over", "expired", "refunded", "paid_remain"} {
		t.Run(eventType, func(t *testing.T) {
			payload := callbackEnvelope()
			payload["event_name"] = "order"
			payload["event_type"] = eventType
			payload["data"] = paymentCallbackData(eventType)
			result, parseErr := handler.VerifyAndParse(signedCallbackBody(t, payload))
			if parseErr != nil {
				t.Fatal(parseErr)
			}
			if string(result.Envelope.EventType) != eventType {
				t.Fatalf("event type = %q, want %q", result.Envelope.EventType, eventType)
			}
			switch eventType {
			case "expired":
				if _, ok := result.Data.(*PaymentExpiredCallbackData); !ok {
					t.Fatalf("unexpected data type %T", result.Data)
				}
			case "refunded":
				if _, ok := result.Data.(*PaymentRefundedCallbackData); !ok {
					t.Fatalf("unexpected data type %T", result.Data)
				}
			case "paid_remain":
				if _, ok := result.Data.(*PaymentRemainingCallbackData); !ok {
					t.Fatalf("unexpected data type %T", result.Data)
				}
			default:
				if _, ok := result.Data.(*PaymentCompletedCallbackData); !ok {
					t.Fatalf("unexpected data type %T", result.Data)
				}
			}
		})
	}
	if CallbackSuccess != "success" {
		t.Fatalf("unexpected callback acknowledgement: %q", CallbackSuccess)
	}
}

func TestAllFourWaaSWebhookContractsVerifyAndParse(t *testing.T) {
	handler, err := NewWaaSCallbackHandler(callbackAPIKey, WithWebhookClock(callbackTestClock))
	if err != nil {
		t.Fatal(err)
	}

	depositPayload := waasCallbackBase()
	depositPayload["address"] = "deposit-address"
	depositPayload["status"] = "1"
	depositPayload["txid"] = "tx-1"
	depositPayload["block_time"] = "1734328473070"
	deposit, err := handler.HandleDepositCallback(signedCallbackBody(t, depositPayload))
	if err != nil || deposit.Status != "1" {
		t.Fatalf("deposit callback failed: value=%+v error=%v", deposit, err)
	}

	payoutPayload := waasCallbackBase()
	payoutPayload["address"] = "payout-address"
	payoutPayload["third_party_id"] = "payout-1"
	payoutPayload["status"] = 6
	payoutPayload["block_time"] = int64(1734328473070)
	payout, err := handler.HandlePayoutCallback(signedCallbackBody(t, payoutPayload))
	if err != nil || payout.Status != 6 {
		t.Fatalf("payout callback failed: value=%+v error=%v", payout, err)
	}

	externalPayload := cloneMap(waasCallbackBase())
	delete(externalPayload, "currency")
	externalPayload["third_party_id"] = "external-1"
	externalPayload["from_address"] = "from"
	externalPayload["to_address"] = "to"
	external, err := handler.HandlePayoutExternalVerificationCallback(signedCallbackBody(t, externalPayload))
	if err != nil || external.ThirdPartyID != "external-1" {
		t.Fatalf("external verification callback failed: value=%+v error=%v", external, err)
	}

	withdrawalPayload := waasCallbackBase()
	withdrawalPayload["from_address"] = "from"
	withdrawalPayload["to_address"] = "to"
	withdrawalPayload["third_party_id"] = "withdrawal-1"
	withdrawalPayload["status"] = 6
	withdrawal, err := handler.HandleWithdrawalCallback(signedCallbackBody(t, withdrawalPayload))
	if err != nil || withdrawal.ToAddress != "to" {
		t.Fatalf("withdrawal callback failed: value=%+v error=%v", withdrawal, err)
	}

	if ExternalVerificationApprove != "ok" || ExternalVerificationDeny != "deny" {
		t.Fatal("unexpected external verification acknowledgement constants")
	}
}

func TestWebhooksRejectBadSignaturesUnknownEventsAndWrongTypes(t *testing.T) {
	payment, err := NewPaymentCallbackHandler(callbackAPIKey, WithWebhookClock(callbackTestClock))
	if err != nil {
		t.Fatal(err)
	}
	waas, err := NewWaaSCallbackHandler(callbackAPIKey, WithWebhookClock(callbackTestClock))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := payment.VerifyAndParse([]byte("{}")); err == nil {
		t.Fatal("expected unsigned callback to fail")
	}

	unknown := callbackEnvelope()
	unknown["event_name"] = "order"
	unknown["event_type"] = "future_event"
	unknown["data"] = map[string]any{}
	_, err = payment.VerifyAndParse(signedCallbackBody(t, unknown))
	var contractError *ContractError
	if !errors.As(err, &contractError) {
		t.Fatalf("expected unknown event ContractError, got %T: %v", err, err)
	}

	wrongType := waasCallbackBase()
	wrongType["address"] = "deposit-address"
	wrongType["status"] = 1
	wrongType["txid"] = "tx-1"
	_, err = waas.HandleDepositCallback(signedCallbackBody(t, wrongType))
	if !errors.As(err, &contractError) {
		t.Fatalf("expected wrong wire type ContractError, got %T: %v", err, err)
	}

	badSignature := signedCallbackBody(t, wrongType)
	var payload map[string]any
	if err := json.Unmarshal(badSignature, &payload); err != nil {
		t.Fatal(err)
	}
	payload["amount"] = "tampered"
	badSignature, err = json.Marshal(payload)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := VerifyProjectWebhook(badSignature, callbackAPIKey); err == nil {
		t.Fatal("expected tampered callback to fail")
	}
}

func TestWebhookVerificationRejectsTimestampProjectAndReplayViolations(t *testing.T) {
	payload := callbackEnvelope()
	body := signedCallbackBody(t, payload)

	if _, err := VerifyProjectWebhook(body, callbackAPIKey,
		WithWebhookClock(func() time.Time { return callbackTestClock().Add(11 * time.Minute) }),
	); err == nil || !strings.Contains(err.Error(), "older") {
		t.Fatalf("expected stale callback rejection, got %v", err)
	}
	if _, err := VerifyProjectWebhook(body, callbackAPIKey,
		WithWebhookClock(func() time.Time { return callbackTestClock().Add(-2 * time.Minute) }),
	); err == nil || !strings.Contains(err.Error(), "future") {
		t.Fatalf("expected future callback rejection, got %v", err)
	}
	if _, err := VerifyProjectWebhook(body, callbackAPIKey,
		WithWebhookClock(callbackTestClock),
		WithExpectedProjectID(1382528827416577),
	); err == nil || !strings.Contains(err.Error(), "expected project") {
		t.Fatalf("expected project mismatch rejection, got %v", err)
	}

	replayed := errors.New("callback already processed")
	seen := false
	guard := WebhookReplayGuardFunc(func(projectID int64, nonce string, timestamp time.Time) error {
		if projectID != 1382528827416576 || nonce != "m8jisx" || !timestamp.Equal(callbackTestClock()) {
			t.Fatalf("unexpected replay identity: pid=%d nonce=%q timestamp=%s", projectID, nonce, timestamp)
		}
		if seen {
			return replayed
		}
		seen = true
		return nil
	})
	options := []WebhookOption{WithWebhookClock(callbackTestClock), WithWebhookReplayGuard(guard)}
	if _, err := VerifyProjectWebhook(body, callbackAPIKey, options...); err != nil {
		t.Fatalf("first callback failed: %v", err)
	}
	if _, err := VerifyProjectWebhook(body, callbackAPIKey, options...); !errors.Is(err, replayed) {
		t.Fatalf("expected replay guard error, got %v", err)
	}
}

func TestPaymentCallbackRejectsInvalidDocumentedStatus(t *testing.T) {
	handler, err := NewPaymentCallbackHandler(callbackAPIKey, WithWebhookClock(callbackTestClock))
	if err != nil {
		t.Fatal(err)
	}
	payload := callbackEnvelope()
	payload["event_name"] = "order"
	payload["event_type"] = "paid"
	data := paymentCallbackData("paid")
	data["status"] = "not-a-payment-status"
	payload["data"] = data
	_, err = handler.VerifyAndParse(signedCallbackBody(t, payload))
	var contractError *ContractError
	if !errors.As(err, &contractError) || !strings.Contains(err.Error(), "status must be one of") {
		t.Fatalf("expected callback enum ContractError, got %T: %v", err, err)
	}
}
